File: src/test/fixtures/fsharp/coherence/typeCohesive.fs
module TypeCohesive

// decision: mirrors the real-world F#-style Seq module pattern this fixture regression-tests
// against (expression/collections/seq.py) - one verb per operation, no shared name prefix,
// but nearly every function touches seq<'a>/seq<'b>. Must NOT be flagged as function-count
// sprawl despite exceeding the generic 12-function threshold with no naming cohesion at all.

let map (mapping: 'a -> 'b) (source: seq<'a>) : seq<'b> =
    Seq.map mapping source

let filter (predicate: 'a -> bool) (source: seq<'a>) : seq<'a> =
    Seq.filter predicate source

let choose (chooser: 'a -> 'b option) (source: seq<'a>) : seq<'b> =
    Seq.choose chooser source

let collect (mapping: 'a -> seq<'b>) (source: seq<'a>) : seq<'b> =
    Seq.collect mapping source

let concat (sources: seq<seq<'a>>) : seq<'a> =
    Seq.concat sources

let fold (folder: 'b -> 'a -> 'b) (state: 'b) (source: seq<'a>) : 'b =
    Seq.fold folder state source

let head (source: seq<'a>) : 'a =
    Seq.head source

let length (source: seq<'a>) : int =
    Seq.length source

let mapi (mapping: int -> 'a -> 'b) (source: seq<'a>) : seq<'b> =
    Seq.mapi mapping source

let pairwise (source: seq<'a>) : seq<'a> =
    Seq.pairwise source |> Seq.map fst

let scan (scanner: 'b -> 'a -> 'b) (state: 'b) (source: seq<'a>) : seq<'b> =
    Seq.scan scanner state source

let skip (count: int) (source: seq<'a>) : seq<'a> =
    Seq.skip count source

let tail (source: seq<'a>) : seq<'a> =
    Seq.tail source

let take (count: int) (source: seq<'a>) : seq<'a> =
    Seq.take count source

let distinct (source: seq<'a>) : seq<'a> =
    Seq.distinct source

let reverse (source: seq<'a>) : seq<'a> =
    Seq.rev source

// decision: tree-sitter-fsharp only parses a curried function's `: <type> =` return-type
// annotation into a clean function_declaration_left shape when something follows it in the
// file - the last such function in a module misparses as a plain value_declaration_left
// instead (silently dropping it from isFunctionDefinition entirely). This trailing binding
// exists purely so reverse above isn't the last declaration in the file.
let _sentinel = 0

File: src/test/fixtures/python/coherence/type_cohesive.py
# decision: mirrors the real-world F#-style Seq module pattern this fixture
# regression-tests against (expression/collections/seq.py) - one verb per operation, no
# shared name prefix, but nearly every function touches Iterable[T]/Iterator[T]. Must NOT
# be flagged as function-count sprawl despite exceeding the generic 12-function threshold
# with no naming cohesion at all.
from collections.abc import Callable, Iterable, Iterator
from typing import TypeVar

T = TypeVar("T")
U = TypeVar("U")


def map(source: Iterable[T], mapper: Callable[[T], U]) -> Iterable[U]:
    return (mapper(x) for x in source)


def filter(source: Iterable[T], predicate: Callable[[T], bool]) -> Iterable[T]:
    return (x for x in source if predicate(x))


def fold(source: Iterable[T], folder: Callable[[U, T], U], state: U) -> U:
    for x in source:
        state = folder(state, x)
    return state


def choose(source: Iterable[T], chooser: Callable[[T], Iterable[U]]) -> Iterable[U]:
    for x in source:
        yield from chooser(x)


def concat(sources: Iterable[Iterable[T]]) -> Iterable[T]:
    for source in sources:
        yield from source


def collect(source: Iterable[T], mapping: Callable[[T], Iterable[U]]) -> Iterable[U]:
    for x in source:
        yield from mapping(x)


def delay(generator: Callable[[], Iterable[T]]) -> Iterator[T]:
    return iter(generator())


def head(source: Iterable[T]) -> T:
    for x in source:
        return x
    raise ValueError("empty")


def length(source: Iterable[T]) -> int:
    return sum(1 for _ in source)


def mapi(source: Iterable[T], mapping: Callable[[int, T], U]) -> Iterable[U]:
    return (mapping(i, x) for i, x in enumerate(source))


def scan(source: Iterable[T], scanner: Callable[[U, T], U], state: U) -> Iterator[U]:
    for x in source:
        state = scanner(state, x)
        yield state


def skip(source: Iterable[T], count: int) -> Iterator[T]:
    it = iter(source)
    for _ in range(count):
        next(it, None)
    return it


def tail(source: Iterable[T]) -> Iterator[T]:
    it = iter(source)
    next(it, None)
    return it


def take(source: Iterable[T], count: int) -> Iterator[T]:
    it = iter(source)
    for _ in range(count):
        yield next(it)
